### ISSUE TITLE
Build error on macos

### ISSUE THREAD

**Issue by ct16k** [jemalloc/jemalloc#1467]
Hello,

Trying to build both the current tip and last release on macOS Mojave fails for me because of background thread bits:
```
src/background_thread.c:881:6: warning: implicit declaration of function 'pthread_create_fptr_init' is invalid in C99 [-Wimplicit-function-declaration]
            pthread_create_fptr_init()) {
            ^
```
I tried with both the built-in compiler, as well as clang 7 and the newly released 8.

Config output is:
```
jemalloc version   : 5.1.0-154-gfb56766ca9b398d07e2def5ead75a021fc08da03
library revision   : 2

CONFIG             : --enable-autogen --enable-xmalloc --enable-lazy-lock --with-malloc-conf=background_thread:true,metadata_thp:auto CC=/usr/local/clang+llvm-8.0.0-x86_64-apple-darwin/bin/clang 'LDFLAGS=-L/usr/local/clang+llvm-8.0.0-x86_64-apple-darwin/lib -Wl,-rpath,/us
r/local/clang+llvm-8.0.0-x86_64-apple-darwin/lib ' 'CPPFLAGS=-I/usr/local/clang+llvm-8.0.0-x86_64-apple-darwin/include -I/usr/local/clang+llvm-8.0.0-x86_64-apple-darwin/include/c++/v1/ ' CXX=/usr/local/clang+llvm-8.0.0-x86_64-apple-darwin/bin/clang++
CC                 : /usr/local/clang+llvm-8.0.0-x86_64-apple-darwin/bin/clang
CONFIGURE_CFLAGS   : -std=gnu11 -Wall -Wextra -Wshorten-64-to-32 -Wsign-compare -Wundef -Wno-format-zero-length -pipe -g3 -O3 -funroll-loops
SPECIFIED_CFLAGS   :
EXTRA_CFLAGS       :
CPPFLAGS           : -I/usr/local/clang+llvm-8.0.0-x86_64-apple-darwin/include -I/usr/local/clang+llvm-8.0.0-x86_64-apple-darwin/include/c++/v1/  -D_REENTRANT
CXX                : /usr/local/clang+llvm-8.0.0-x86_64-apple-darwin/bin/clang++
CONFIGURE_CXXFLAGS : -Wall -Wextra -g3 -O3
SPECIFIED_CXXFLAGS :
EXTRA_CXXFLAGS     :
LDFLAGS            : -L/usr/local/clang+llvm-8.0.0-x86_64-apple-darwin/lib -Wl,-rpath,/usr/local/clang+llvm-8.0.0-x86_64-apple-darwin/lib
EXTRA_LDFLAGS      :
DSO_LDFLAGS        : -shared -Wl,-install_name,$(LIBDIR)/$(@F)
LIBS               : -lstdc++ -pthread
RPATH_EXTRA        :

XSLTPROC           : /usr/bin/xsltproc
XSLROOT            :

PREFIX             : /usr/local
BINDIR             : /usr/local/bin
DATADIR            : /usr/local/share
INCLUDEDIR         : /usr/local/include
LIBDIR             : /usr/local/lib
MANDIR             : /usr/local/share/man

srcroot            :
abs_srcroot        : /usr/local/jemalloc/
objroot            :
abs_objroot        : /usr/local/jemalloc/

JEMALLOC_PREFIX    : je_
JEMALLOC_PRIVATE_NAMESPACE
                   : je_
install_suffix     :
malloc_conf        : background_thread:true,metadata_thp:auto
shared libs        : 1
static libs        : 1
autogen            : 1
debug              : 0
stats              : 1
experimetal_smallocx : 0
prof               : 0
prof-libunwind     : 0
prof-libgcc        : 0
prof-gcc           : 0
fill               : 1
utrace             : 0
xmalloc            : 1
log                : 0
lazy_lock          : 1
cache-oblivious    : 1
cxx                : 1
```
I found that manually defining `JEMALLOC_BACKGROUND_THREAD` through `CFLAGS` gets me over this, but I'm not sure if it's the proper way to handle this.

**Comment by interwq** [comment]
It's likely because of the config option: `malloc_conf        : background_thread:true`. Currently the `background_thread` feature depends on pthread support, which we haven't fully migrated to on MacOS yet. See #843.

Closing but feel free to ask if anything.

**Comment by ct16k** [comment]
Even without `background_thread:true` I still get the same failure, and I guess that would be expected, since it is my understanding that that sets default runtime config, while I'm having issues compiling the library (I was setting it to true since I just used the same params I did on Linux).

Background threads seem to indeed be unsupported on this version of macOS, given that they're considered available only when pthread is present but os_unfair_lock is not, which isn't the case. The issue here is that `pthread_create_fptr_init` declaration is guarded by `#ifdef JEMALLOC_BACKGROUND_THREAD`, then used outside of it, in `background_thread_boot0`. That's why in my previous comment I mentioned I was able to get it to build only by manually defining `JEMALLOC_BACKGROUND_THREAD`. Though I don't know if this is OK to do as long as I'm careful to not actually enable background threads, or if a better approach would be to change the `#ifdef` check in `background_thread_boot0` for `JEMALLOC_BACKGROUND_THREAD` instead of `JEMALLOC_PTHREAD_CREATE_WRAPPER`, to avoid the call to `pthread_create_fptr_init`. Or if the fix is not at all that trivial.
